### PR TITLE
Remove MutexID list

### PR DESCRIPTION
**Pull request by tiif**
Remove MutexID list in ``SynchronizationObjects`` to avoid memory leaks. 

cc #3967

### CODE REVIEW

File: tests/pass-dep/libc/pthread-sync.rs [hunk 1]
```diff
@@ -215,6 +214,8 @@ fn check_rwlock_read_no_deadlock() {
     }
 }
 
+// TODO: There is a data race error when there is no -Zmiri-preemption-rate=0, test this again after
+// the ICE is resolved.
 fn check_cond() {
```
Comment (tiif) [inline tests/pass-dep/libc/pthread-sync.rs:219]: Surprisingly miri reported data race on this test when there is no ``-Zmiri-preemption-rate=0``. I haven't looked into this yet, but I will revert this change for now as ``-Zmiri-preemption-rate=0`` is needed to trigger the ICE.

**Comment by tiif** [comment]
I feel I have been constantly running into ``already mutably borrowed: BorrowError`` when it comes to ``unblock_thread``. It always takes me a while to realise that it has something to do with ``unblock_thread`` again.

**Comment by RalfJung** [comment]
Generally you have to take great care of your `RefCell` handles. Never keep them any longer than you have to.

If we didn't use `RefCell`, you'd get a borrow checking error due to changing a vector while iterating it. That would be nicer but unfortunately the reference patterns are too dynamic to make that work.

**Comment by tiif** [comment]
> For every function you call while holding the handle, ensure it doesn't reentrantly access the same data again.

This is a pretty nice way to see it, thanks.

File: tests/pass-dep/libc/pthread-sync.rs (hunk 1 above)
Comment (tiif) [reply tests/pass-dep/libc/pthread-sync.rs:220]: Before the data race is reported, there is another failed assertion reported, so the data race might be an side effect of the failed assertion? If no one's alarm went off about this, it's probably fine. 
<details>
<summary> error trace </summary>

```rust
thread 'main' panicked at ./tests/pass-dep/libc/pthread-sync.rs:247:9:
assertion failed: data.ptr.read() == 1
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
note: in Miri, you may have to set `MIRIFLAGS=-Zmiri-env-forward=RUST_BACKTRACE` for the environment variable to have an effect
error: Undefined Behavior: Data race detected between (1) atomic store on thread `unnamed-11` and (2) deallocation on thread `main` at alloc23901. (2) just happened here
   --> ./tests/pass-dep/libc/pthread-sync.rs:218:1
    |
218 | / fn check_cond() {
219 | |     unsafe {
220 | |         let mut cond: MaybeUninit<libc::pthread_cond_t> = MaybeUninit::uninit();
221 | |         assert_eq!(libc::pthread_cond_init(cond.as_mut_ptr(), ptr::null()), 0);
...   |
263 | |     }
264 | | }
    | |_^ Data race detected between (1) atomic store on thread `unnamed-11` and (2) deallocation on thread `main` at alloc23901. (2) just happened here
    |
help: and (1) occurred earlier here
   --> ./tests/pass-dep/libc/pthread-sync.rs:234:24
    |
234 |             assert_eq!(libc::pthread_mutex_lock(mutex.ptr), 0); // This is one
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
    = note: BACKTRACE (of the first span):
    = note: inside `check_cond` at ./tests/pass-dep/libc/pthread-sync.rs:218:1: 264:2
note: inside `main`
   --> ./tests/pass-dep/libc/pthread-sync.rs:21:5
    |
21  |     check_cond();
    |     ^^^^^^^^^^^^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error

error: test failed, to rerun pass `--test ui`

```
</details>

But I couldn't understand the diagnostic. ``(2)`` is pointed at the end of the function, which is not really helpful for me to understand where exactly it is. 

```rust
218 | / fn check_cond() {
219 | |     unsafe {
220 | |         let mut cond: MaybeUninit<libc::pthread_cond_t> = MaybeUninit::uninit();
221 | |         assert_eq!(libc::pthread_cond_init(cond.as_mut_ptr(), ptr::null()), 0);
...   |
263 | |     }
264 | | }
    | |_^ Data race detected between (1) atomic store on thread `unnamed-11` and (2) deallocation on thread `main` at alloc23901. (2) just happened here
```

File: src/concurrency/sync.rs [hunk 2]
```diff
@@ -44,7 +47,20 @@ macro_rules! declare_id {
...
+    fn new() -> Self {
+        MutexRef(Rc::new(RefCell::new(Mutex::default())))
+    }
+}
+
+impl VisitProvenance for MutexRef {
+    fn visit_provenance(&self, _visit: &mut VisitWith<'_>) {
+        // No provenance
+    }
+}
```
Comment (tiif) [inline src/concurrency/sync.rs:63]: Is there anything need to be implemented here?

File: src/shims/unix/sync.rs [hunk 3]
```diff
@@ -176,10 +176,11 @@ fn mutex_create<'tcx>(
...
     let data = PthreadMutex { id, kind };
-    ecx.lazy_sync_init(&mutex, mutex_init_offset(ecx)?, data)?;
+    ecx.lazy_sync_init(&mutex, mutex_init_offset(ecx)?, data.clone())?;
     interp_ok(data)
 }
 
+// TODO: check if the comment is correct?
 /// Returns the `MutexId` of the mutex stored at `mutex_op`.
 ///
 /// `mutex_get_id` will also check if the mutex has been moved since its first use and
```
Comment (tiif) [inline src/shims/unix/sync.rs:186]: This comment looks a bit weird. ``mutex_get_data`` returns a ``PthreadMutex``, and why is there ``mutex_op`` and ``mutex_get_id``?

**Review by tiif — COMMENTED** [review]
Aside from the questions in the comments, this is ready

@rustbot ready

File: tests/pass-dep/libc/pthread-sync.rs (hunk 1 above)
Comment (RalfJung) [reply tests/pass-dep/libc/pthread-sync.rs]: What is the status of this? The diff seems to be done, there is no ICE.

> (2) is pointed at the end of the function

Yes, that's where the local variables get deallocated.

Comment (RalfJung) [reply tests/pass-dep/libc/pthread-sync.rs]: The reason for the race is that we are panicking, and then we are not executing the `t.join().unwrap()`, and then we get a race.

Comment (RalfJung) [reply tests/pass-dep/libc/pthread-sync.rs]: Resolved by https://github.com/rust-lang/miri/pull/4020

File: src/concurrency/sync.rs [hunk 4]
```diff
@@ -44,7 +47,20 @@ macro_rules! declare_id {
 }
 pub(super) use declare_id;
 
-declare_id!(MutexId);
+#[derive(Default, Clone, Debug)]
+pub struct MutexRef(Rc<RefCell<Mutex>>);
```
Comment (RalfJung) [inline src/concurrency/sync.rs:51]: Please define this right after `struct Mutex` so that related things are close together.

File: src/concurrency/sync.rs (hunk 2 above)
Comment (RalfJung) [reply src/concurrency/sync.rs:63]: No, please add a comment saying that `Mutex` contains no provenance.

File: src/concurrency/sync.rs [hunk 5]
```diff
@@ -147,17 +162,17 @@ impl<'tcx> EvalContextExtPriv<'tcx> for crate::MiriInterpCx<'tcx> {}
 pub(super) trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
     fn condvar_reacquire_mutex(
         &mut self,
-        mutex: MutexId,
+        mutex_ref: MutexRef,
```
Comment (RalfJung) [inline src/concurrency/sync.rs:165]: This should take `&MutexRef`, no need to make the caller give up ownership here

File: src/concurrency/sync.rs [hunk 6]
```diff
@@ -354,15 +366,18 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                         mutex.clock.clone_from(clock)
                     });
                 }
-                if let Some(thread) = this.machine.sync.mutexes[id].queue.pop_front() {
-                    this.unblock_thread(thread, BlockReason::Mutex(id))?;
+                let thread_id = mutex.queue.pop_front();
+                if thread_id.is_some() {
+                    // We need to drop mutex before unblock_thread because mutex will be borrowed again
+                    // in the unblock callback.
```
Comment (RalfJung) [inline src/concurrency/sync.rs:372]:
```suggestion
                    // We need to drop mutex before unblock_thread because mutex will be borrowed again
                    // in the unblock callback.
```
```suggestion
                    // We need to drop our mutex borrow before unblock_thread
                    //  because it will be borrowed again in the unblock callback.
```

File: src/concurrency/sync.rs [hunk 7]
```diff
@@ -354,15 +366,18 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+                    drop(mutex);
```
Comment (RalfJung) [inline src/concurrency/sync.rs:373]: Can't you do the drop before the if? That seems cleaner. Always drop as early as possible.

File: src/concurrency/sync.rs [hunk 8]
```diff
@@ -354,15 +366,18 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+                    this.unblock_thread(thread_id.unwrap(), BlockReason::Mutex)?;
                 }
             }
-            Some(old_lock_count)
-        } else {
-            // Mutex is not locked.
-            None
-        })
+            return interp_ok(Some(old_lock_count));
```
Comment (RalfJung) [inline src/concurrency/sync.rs:377]: Why did you change this to use early-return?

File: src/concurrency/sync.rs [hunk 9]
```diff
@@ -227,7 +241,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
     /// - If yes, fetches the data from `alloc_extra.sync`, or calls `missing_data` if that fails
     ///   and stores that in `alloc_extra.sync`.
     /// - Otherwise, calls `new_data` to initialize the primitive.
```
Comment (RalfJung) [inline src/concurrency/sync.rs:243]:
```suggestion
    /// - Otherwise, calls `new_data` to initialize the primitive.
    ///
    /// The return value is a *clone* of the stored data, so if you intend to mutate it
    /// better wrap everything into an `Rc`.
```

File: src/shims/unix/sync.rs (hunk 3 above)
Comment (RalfJung) [reply src/shims/unix/sync.rs:186]: Yeah that comment is just outdated. Make it like this:
```
/// Returns the mutex data stored at the address that `mutex_ptr` points to.
/// Will raise an error if the mutex has been move since its first use.
```

**Review by RalfJung — CHANGES_REQUESTED** [review]
Thanks, looks great overall! But there's a bunch of minor things.

File: src/concurrency/sync.rs (hunk 8 above)
Comment (tiif) [reply src/concurrency/sync.rs]: That's a spurious change introduced when I try to debug fix the reborrow ICE. It is now fixed.

File: src/concurrency/sync.rs [hunk 10]
```diff
@@ -610,14 +630,14 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
     fn condvar_wait(
         &mut self,
         condvar: CondvarId,
-        mutex: MutexId,
+        mutex_ref: &MutexRef,
```
Comment (RalfJung) [inline src/concurrency/sync.rs:633]: This one needs to clone anyway so it makes sense that it would take an owned ref (so we can avoid double-cloning).

File: src/shims/unix/sync.rs [hunk 11]
```diff
@@ -175,15 +175,13 @@ fn mutex_create<'tcx>(
...
+    ecx.lazy_sync_init(&mutex, mutex_init_offset(ecx)?, data.clone())?;
     interp_ok(data)
 }
 
-/// Returns the `MutexId` of the mutex stored at `mutex_op`.
-///
-/// `mutex_get_id` will also check if the mutex has been moved since its first use and
-/// return an error if it has.
+/// Returns the mutex data stored at the address that `mutex_ptr` points to.
+/// Will raise an error if the mutex has been move since its first use.
```
Comment (RalfJung) [inline src/shims/unix/sync.rs:184]:
```suggestion
/// Will raise an error if the mutex has been moved since its first use.
```

File: src/concurrency/sync.rs [hunk 12]
```diff
@@ -354,8 +369,12 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                         mutex.clock.clone_from(clock)
                     });
                 }
-                if let Some(thread) = this.machine.sync.mutexes[id].queue.pop_front() {
-                    this.unblock_thread(thread, BlockReason::Mutex(id))?;
+                let thread_id = mutex.queue.pop_front();
+                // We need to drop our mutex borrow before unblock_thread
+                //  because it will be borrowed again in the unblock callback.
```
Comment (RalfJung) [inline src/concurrency/sync.rs:374]:
```suggestion
                // because it will be borrowed again in the unblock callback.
```

**Comment by RalfJung** [comment]
Just some final nits. :)
You can squash after taking care of those (with `--keep-base` as usual).

@rustbot author

**Review by RalfJung — APPROVED** [review]

**Comment by RalfJung** [comment]
Oh, wtf... as long as there is a "requested changes" review, it won't merge, even if the merge button is pressed by the person that requested the changes...
